### PR TITLE
Support this cast to base class.

### PR DESCRIPTION
**Pull request by python3kgae**
Copy will be generated from this to tmp base class ptr.

### CODE REVIEW

**Comment by llvm-beanz** [comment]
Should this code be valid? In C++ implicit cast-to-base is only valid for pointers not for value types.

**Comment by python3kgae** [comment]
> Should this code be valid? In C++ implicit cast-to-base is only valid for pointers not for value types.

I thought 'this' is a "reference" for hlsl. :(

**Comment by llvm-beanz** [comment]
> I thought 'this' is a "reference" for hlsl. :(

Sure, but the parameter is a value parameter. So it gets copied.

**Comment by llvm-beanz** [comment]
FWIW, under HLSL 2018, the code in this test case is probably valid because of the old UDT casting rules that allowed structs of the same layout to be implicitly casted. It would not be valid under HLSL 2021+.

**Comment by mstrgram** [comment]
If `this` is to be considered a value type, I assume all objects passed to methods are to have value semantics. If that is the case, shouldn't something like below also fail to compile then?

``` 
struct A
{
    uint _member;
    
    bool SomeFunc(A other)
    {
        return _member == other._member;
    }
};

struct B : A
{
    uint _extraMember;
};

A a = { 10 };
B b = { 11, 12 };
bool same = a.SomeFunc(b);
```

**Comment by python3kgae** [comment]
> FWIW, under HLSL 2018, the code in this test case is probably valid because of the old UDT casting rules that allowed structs of the same layout to be implicitly casted. It would not be valid under HLSL 2021+.

I thought HLSL 2021 is closer to C++.

But C++ is fine for code like this where reference used for a value parameter.
```
struct A {
    float a;
};

struct B : public A {
   int b;
};


float foo(A a) {
    return a.a;
}

float bar(B &b) {
    return foo(b);
}
```

**Comment by llvm-beanz** [comment]
> But C++ is fine for code like this where reference used for a value parameter.
> 
> ```
> struct A {
>     float a;
> };
> 
> struct B : public A {
>    int b;
> };
> 
> 
> float foo(A a) {
>     return a.a;
> }
> 
> float bar(B &b) {
>     return foo(b);
> }
> ```

The way this works in C++ is slightly unintuitive. Because `A` has implicit copy constructors what ends up happening is that a new `A` gets constructed from a `B&` casted up to an `A&`. This only works because the constructors are present.

If you add `A(const A&) = delete;` to `A`, the code fails.

Since we don't have constructors in HLSL, this code only works because we've defined a truncating cast 🤮 .

**Comment by mstrgram** [comment]
> We should add some additional tests to verify parameter passing semantics are correctly observed.
>
> Specifically copy-in/copy-out for this objects implicitly casted to bases. We should make sure that the truncated base object is what is copied in all cases when an HLSLDerivedToBase cast is present.

Does this mean you agree the original `this` cast to base and the implicit `B` --> `A` param cast I showed should be valid?

**Comment by llvm-beanz** [comment]
> Does this mean you agree the original `this` cast to base and the implicit `B` --> `A` param cast I showed should be valid?

Yes, but keep in mind this is a truncating value cast. When used in a parameter it amounts to copying the base class data out of the derived object.

When used with `inout` it should result in copying the base data out, then copying it back at the end of the function, and when used with an `out` parameter it would result in only initializing the base class portion of `this`.

**Review by llvm-beanz — COMMENTED** [review]
Your test cases here are revealing that we're not eliding copy-in semantics in some places we could safely do so. That said, we can resolve that separately. This code generation is sub-optimal but correct.

A few minor notes inline.

File: tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl [hunk 1]
```diff
@@ -1,5 +1,36 @@
...
+// CHECK: %[[OutArgCpyPtr:.+]] = bitcast %class.Parent* %[[OutArg]] to i8*
+// CHECK: call void @llvm.memcpy.p0i8.p0i8.i64(i8* %[[OutThisCpyPtr]], i8* %[[OutArgCpyPtr]], i64 8, i32 1, i1 false)
+
+// CHECK-LABLE: define linkonce_odr void @"\01?bar@Child@@QAAXXZ"(%class.Child* %this)
+// CHECK: %[[InOutArg:.+]] = alloca %class.Parent
+// CHECK: %[[InOutThisPtr:.+]] = getelementptr inbounds %class.Child, %class.Child* %this, i32 0, i32 0
+
+// Skip lifetime marker
+// CHECK: %[[InOutArgPtrForMarker:.+]] = bitcast %class.Parent* %[[InOutArg]] to i8*
+// CHECK: call void @llvm.lifetime.start(i64 8, i8* %[[InOutArgPtrForMarker]])
```
[the remark this answers is not in the document]
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl:18]: It will make an extra bitcast, one solution could be disable lifetime marker.

File: tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl [hunk 2]
```diff
@@ -33,13 +64,21 @@ class Parent
 };
 
 int lib_func(Parent obj, int a, int b);
+void lib_func2(out Parent obj);
+void lib_func3(inout Parent obj);
 
 class Child : Parent
 {
     int foo(int a, int b)
```
[the remark this answers is not in the document]
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl:64]: Done.

File: tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl [hunk 3]
```diff
@@ -0,0 +1,93 @@
+// RUN: %dxc -T lib_6_6 -HV 2021 -disable-lifetime-markers -fcgl %s | FileCheck %s
```
Comment (llvm-beanz) [inline tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl:1]: Lifetime markers shouldn’t impact the test. If it does, your test matches are too fragile. You should be able to use a `CHECK-DAG` to work around the `bitcast` instructions inserted with lifetime markers.

File: tools/clang/lib/Sema/SemaOverload.cpp [hunk 4]
```diff
@@ -4967,8 +4967,7 @@ InitCallParamConversions(Sema &S, const FunctionProtoType *Proto,
     Expr *OutFrom = DeclRefExpr::Create(
         S.getASTContext(), NestedNameSpecifierLoc(), SourceLocation(), Param,
         true, Param->getLocation(), ParamType.getNonReferenceType(), VK_RValue, nullptr);
-    OutConversion = TryCopyInitialization(
-        S, OutFrom, Arg->getType(), SuppressUserConversions,
+    OutConversion = TryCopyInitialization(S, OutFrom, ParamType, SuppressUserConversions,
```
Comment (llvm-beanz) [inline tools/clang/lib/Sema/SemaOverload.cpp:4970]: I’m not sure this change is correct. This alters the AST so that instead of converting the return value from Parent->Child, it converts Parent->Parent.

Comment (llvm-beanz) [reply tools/clang/lib/Sema/SemaOverload.cpp:4970]: It also looks like we don't support `HLSLDerivedToBase` truncation for `inout` or `out` on other parameters (https://godbolt.org/z/aqaKK1qTd), so we don't need to for `this` either.

Comment (python3kgae) [reply tools/clang/lib/Sema/SemaOverload.cpp]: Reverted.

Comment (mstrgram) [reply tools/clang/lib/Sema/SemaOverload.cpp]: Isn't it somewhat strange to have different cast behavior depending on whether the var is `in` only or `inout`/ `out`?

Comment (python3kgae) [reply tools/clang/lib/Sema/SemaOverload.cpp]: Yeah, it is strange.
Do you want to always require cast or always implicit cast for case like this?

Comment (llvm-beanz) [reply tools/clang/lib/Sema/SemaOverload.cpp]: This code is wonky, but I think correct.

For `in` and `inout` parameters we copy initialize a new object from the argument type to the parameter type.

Then for `out` and `inout` parameters we copy initialize from the parameter to the argument type.

The trick here is that `InConversion` and `OutConversion` are two different casts, and they can both be set.

This should correctly model the semantics that `in` and `inout` parameters need to be casted going in, and that `out` and `inout` need to be casted coming out.

Comment (python3kgae) [reply tools/clang/lib/Sema/SemaOverload.cpp]: What's the point to cast to bigger type when the copy has to be truncated?

If cast to parameter type, the copy will be on smaller type which is just normal copy.

Comment (llvm-beanz) [reply tools/clang/lib/Sema/SemaOverload.cpp]: It isn't really a cast. It is copy-initializing.

On `in` and `inout` parameters the argument (value being passed in) is copy-initialized to a temporary of the parameter type, which is passed into the function. For `out` parameters the temporary is uninitialized.
On `out` and `inout` parameters after the call, the temporary value of the parameter-type is copy-initialized back to the argument variable.

So the type conversion go from argument -> parameter on input and from parameter -> argument on output, but these are copy-initializations not casts.

File: tools/clang/lib/Sema/SemaOverload.cpp [hunk 5]
```diff
@@ -4966,7 +4966,8 @@ InitCallParamConversions(Sema &S, const FunctionProtoType *Proto,
     }
     Expr *OutFrom = DeclRefExpr::Create(
         S.getASTContext(), NestedNameSpecifierLoc(), SourceLocation(), Param,
-        true, Param->getLocation(), ParamType.getNonReferenceType(), VK_RValue, nullptr);
+        true, Param->getLocation(), ParamType.getNonReferenceType(), VK_RValue,
+        nullptr);
```
Comment (llvm-beanz) [inline tools/clang/lib/Sema/SemaOverload.cpp:4970]: Seems odd to have a formatting change to unrelated code here.

File: tools/clang/lib/CodeGen/CGExprAgg.cpp [hunk 6]
```diff
@@ -741,10 +741,13 @@ void AggExprEmitter::VisitCastExpr(CastExpr *E) {
         LValue LV;
         if (DeclRefExpr *SrcDecl = dyn_cast<DeclRefExpr>(Src))
           LV = CGF.EmitLValue(SrcDecl);
-        else if (ArraySubscriptExpr *ArraySubExpr = dyn_cast<ArraySubscriptExpr>(Src))
+        else if (ArraySubscriptExpr *ArraySubExpr =
+                     dyn_cast<ArraySubscriptExpr>(Src))
```
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGExprAgg.cpp:745]: Is this formatting-only change on purpose?

File: tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+void lib_func3(inout Parent obj);
+
+class Child : Parent
+{
+    int foo(int a, int b)
+    {
+        return lib_func(this, a, b);
+    }
+    void foo() {
+        lib_func2((Parent)this);
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl:69]: Is explicit cast here required for `out` param?  Same for `inout` case in lib_func3 call?

File: tools/clang/lib/CodeGen/CGExprAgg.cpp (hunk 6 above)
Comment (python3kgae) [reply tools/clang/lib/CodeGen/CGExprAgg.cpp:745]: No. I'll revert it.

File: tools/clang/lib/Sema/SemaOverload.cpp (hunk 5 above)
Comment (python3kgae) [reply tools/clang/lib/Sema/SemaOverload.cpp:4970]: Will fix it.

File: tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl (hunk 7 above)
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl:69]: This follows fxc which require explicit cast here.
We could change it in TryCopyInitialization if it is OK to allow implicit cast.

Comment (llvm-beanz) [reply tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl:69]: Ooof... These casts probably cause an mis-compile.

Unless there is some magic and unexplainable casting rule applying here, in these cases the value getting passed in _should_ be the rvalue produced by the cast. They shouldn't result in modifying the underlying value, and we _should_ error on passing rvalues into `out` and `inout` parameters.

Comment (llvm-beanz) [reply tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl:69]: Okay... another sad face from me. This does _seem_ to be the language rules:

https://godbolt.org/z/z47Mqh4fa

Basically, a c-style derived-to-base cast is defined to be some special intuited magic.

File: tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl (hunk 3 above)
Comment (llvm-beanz) [reply tools/clang/test/HLSLFileCheck/hlsl/classes/this_cast_to_base_class.hlsl:1]: I looked at this with @python3kgae. Because the lifetime markers generate identical instructions to the instruction we need to match there isn't a good way to match around it. Disabling lifetime markers simplifies the test, so that's probably the best approach here.

**Review by llvm-beanz — COMMENTED** [review]
I think once the formatting fixes are in, this is good.

**Review by tex3d — APPROVED** [review]
